examples/ul2/src/mod_print_callback.py

```diff
@@ -0,0 +1,214 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+from typing import Any, Dict, List, Optional
+
+import torch
+import transformers
+from composer import Callback, Logger, State
+from composer.utils import dist
+from transformers import PreTrainedModel
+
+from examples.ul2.src import utils
+from examples.ul2.src.data_denoising import MixtureOfDenoisersCollator
+
+__all__ = ['MixtureOfDenoisersPrinterCallback']
+
+
+class MixtureOfDenoisersPrinterCallback(Callback):
+    """Periodically prints model predictions on mixture-of-denoisers examples.
+
+    TODO(Alex): Complete docstring
+    """
+
+    def __init__(
+        self,
+        tokenizer_name: str = 't5-base',
+        print_frequency: int = 500,
+        max_length: int = 128,
+        raise_on_failure: bool = False,
+        generate_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> None:
+
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(
+            tokenizer_name)
+        utils.adapt_tokenizer_for_denoising(self.tokenizer,
+                                            num_sentinel_tokens=100)
+
+        self.print_frequency = int(print_frequency)
+        assert self.print_frequency >= 1
+        self._steps_since_last_print = print_frequency  # This will cause it to print on the first batch, will be reset to 0 each print.
+
+        self.max_length = int(max_length)
+        assert self.max_length >= 1
+
+        self._raise_on_error = bool(raise_on_failure)
+
+        self._generate_kwargs = generate_kwargs if generate_kwargs else {}
+
+        self._verified_collator = False
+        self._decoder_only_format = None
+
+    def init(self, state: State, logger: Logger) -> None:
+        """Verifies model class."""
+        del logger  # unused
+
+        assert hasattr(
+            state.model, 'model'
+        ), 'This callback requires that state.model is a HuggingFaceWrapper instance.'
+        hf_model = state.model.model
+        assert isinstance(
+            hf_model, PreTrainedModel
+        ), 'This class only works for models that subclass `transformers.PreTrainedModel`.'
```

Comment:
we should raise Exception subclasses rather than `assert`-ing when we think the case might actually happen and we aren't in a test. Maybe a TypeError here?

Also, it would be better to [test for behavior, not implementation](https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html). What interface do we need from these objects? Whatever it is, we should ideally just test for that at the point where it's used.

Though if we require a huge interface that only this class would ever implement, or actually _are_ leaning on implementation details, that could be an exception. (These both suggest some pretty tight coupling though...)

Reply:
Noted on the exceptions versus asserts. I will change that.

This callback depends on the HuggingFace `generate` method that comes with `PreTrainedModel`. It's not an interface other classes are going to have, but I will be happy to loosen the check if/when that changes.

I do think it makes sense to keep these checks here (as opposed to where the model is used) to surface any errors as early as possible. It would be a huge pain if this only threw an error after training had already made progress.